components/tidb_query_datatype/src/codec/row/v2/row_slice.rs

```diff
@@ -138,44 +140,106 @@ impl RowSlice<'_> {
 /// This method is only implemented on little endianness currently, since x86 use little endianness.
 #[cfg(target_endian = "little")]
 #[inline]
-fn read_ints_le<'a, T>(buf: &mut &'a [u8], len: usize) -> Result<&'a [T]>
+fn read_le_bytes<'a, T>(buf: &mut &'a [u8], len: usize) -> Result<LEBytes<'a, T>>
 where
     T: PrimInt,
 {
     let bytes_len = std::mem::size_of::<T>() * len;
     if buf.len() < bytes_len {
         return Err(Error::unexpected_eof());
     }
-    let slice = unsafe { std::slice::from_raw_parts(buf.as_ptr() as *const T, len) };
+    let slice = &buf[..bytes_len];
     buf.advance(bytes_len);
-    Ok(slice)
+    Ok(LEBytes::new(slice))
+}
+
+#[cfg(target_endian = "little")]
+pub struct LEBytes<'a, T: PrimInt> {
+    slice: &'a [u8],
+    _marker: PhantomData<T>,
+}
+
+#[cfg(target_endian = "little")]
+impl<'a, T: PrimInt> LEBytes<'a, T> {
+    fn new(slice: &'a [u8]) -> Self {
+        Self {
+            slice,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    #[inline]
+    fn get(&self, index: usize) -> Option<T> {
+        if std::mem::size_of::<T>() * index >= self.slice.len() {
+            None
+        } else {
+            Some(self.get_unchecked(index))
+        }
+    }
+
+    #[inline]
+    fn get_unchecked(&self, index: usize) -> T {
+        let ptr = self.slice.as_ptr() as *const T;
+        unsafe {
+            let ptr = ptr.add(index);
+            std::ptr::read_unaligned(ptr)
+        }
+    }
+
+    #[inline]
+    fn binary_search(&self, value: &T) -> std::result::Result<usize, usize> {
+        let mut size = self.slice.len() / std::mem::size_of::<T>();
+        if size == 0 {
+            return Err(0);
+        }
+        let mut base = 0usize;
+
+        // Note that the count of ids is not greater than (2 >> 16). The number
+        // of binary search steps will not over 16 unless the data is corruted.
```

Comment:
supply some fuzz test? also there is a typo in word "corruted"

Reply:
Haven't. I'll try it tomorrow. (I also reversed the shift left XD)

Reply:
Fuzz test works well. PTAL